L1Trigger/TrackFindingTMTT/src/SimpleLR.cc

```diff
@@ -0,0 +1,495 @@
+        cout << " Stub Residual " << Res << " TP " << const_cast<Stub*>(stub)->assocTP()->index() << endl;
+      else
+        cout << " Stub Residual " << Res << " TP nullptr" << endl;
+    }
+  }
+
+  double LargResidual = 9999.;
+  // Find largest residuals
+  while (vRes.size() > 4 and LargResidual > settings_->ResidualCut()) {
+    std::vector<std::pair<const Stub*, double> >::iterator maxResIt =
+        max_element(vRes.begin(), vRes.end(), pair_compare);
+    LargResidual = (*maxResIt).second;
+    if (settings_->debug() == 6)
+      cout << "Largest Residual " << LargResidual << endl;
+
+    if (LargResidual > settings_->ResidualCut()) {
+      if ((*maxResIt).first->psModule()) {
+        if (psStubs > 2) {
+          if (settings_->debug() == 6)
+            cout << "removing PS residual " << (*maxResIt).second << endl;
+          vRes.erase(maxResIt);
+          psStubs--;
+        } else {
+          if (settings_->debug() == 6)
+            cout << "residual " << (*maxResIt).second << " set to -1. " << endl;
+          (*maxResIt).second = -1.;
+        }
+      } else {
+        vRes.erase(maxResIt);
+        if (settings_->debug() == 6)
+          cout << "removing residual " << (*maxResIt).second << endl;
+      }
+    }
+  }
+
+  std::vector<const Stub*> fitStubs;
+  for (std::pair<const Stub*, double> ResStubPair : vRes) {
+    fitStubs.push_back(ResStubPair.first);
+  }
+
+  phiT = 0.;
+  zT = 0.;
+
+  SumRPhi = 0.;
+  SumR = 0.;
+  SumPhi = 0.;
+  SumR2 = 0.;
+  SumRZ = 0.;
+  SumZ = 0.;
+  double SumR_ps = 0.;
+  double SumR2_ps = 0.;
+
+  numStubs = 0;
+  psStubs = 0;
+
+  for (const Stub* stub : fitStubs) {
+    if ((const_cast<Stub*>(stub))->psModule())
+      psStubs++;
+
+    numStubs++;
+    if (digitize_) {
+      (const_cast<Stub*>(stub))->digitizeForHTinput(l1track3D.iPhiSec());
+      // (const_cast<Stub*>(stub))->digitizeForSFinput();
+      const DigitalStub digiStub = (const_cast<Stub*>(stub))->digitalStub();
+      SumRPhi += digiStub.rt() * digiStub.phiS();
+      SumR += digiStub.rt();
+      SumPhi += digiStub.phiS();
+      SumR2 += digiStub.rt() * digiStub.rt();
+      if ((const_cast<Stub*>(stub))->psModule()) {
+        SumRZ += digiStub.rt() * digiStub.z();
+        SumZ += digiStub.z();
+        SumR_ps += digiStub.rt();
+        SumR2_ps += digiStub.rt() * digiStub.rt();
+      }
+      if (settings_->debug() == 6) {
+        cout << "phiS " << digiStub.iDigi_PhiS() << " rT " << digiStub.iDigi_Rt() << " z " << digiStub.iDigi_Z()
+             << endl;
+      }
+    } else {
+      float phi = 0;
+      if (l1track3D.iPhiSec() == 0 and (const_cast<Stub*>(stub))->phi() > 0) {
+        phi = (const_cast<Stub*>(stub))->phi() - 2 * M_PI;
+      } else if (l1track3D.iPhiSec() == settings_->numPhiSectors() and (const_cast<Stub*>(stub))->phi() < 0) {
+        phi = (const_cast<Stub*>(stub))->phi() + 2 * M_PI;
+      } else {
+        phi = (const_cast<Stub*>(stub))->phi();
```

Comment:
Please avoid `const_cast` (5 cases above and many below, I stop now noting each of them for this function)

Reply:
done